File: WorldFeed/Server/src/WorldFeed.Common/Iterators/MultiLevelCursorIteratorFactory.cs
namespace WorldFeed.Common.Iterators
{
    using System;
    using System.Linq;
    using System.Threading.Tasks;

    using WorldFeed.Common.Iterators.Models;
    using WorldFeed.Common.Public;
    using WorldFeed.Common.Public.Iterators;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Models.Interfaces.DTO.QueryDTO;
    using WorldFeed.Common.Web;

    public interface IMultiLevelCursorIteratorFactory
    {
        IMultiLevelCursorIterator<TInput, TOutput> Create<TDTO, TInput, TOutput>(
            ITwitterPageIterator<ITwitterResult<TDTO>> pageIterator,
            Func<TDTO, TInput[]> transform,
            Func<TInput[], Task<TOutput[]>> getChildItems,
            int maxPageSize);

        IMultiLevelCursorIterator<long, IUser> CreateUserMultiLevelIterator(
            ITwitterClient client,
            ITwitterPageIterator<ITwitterResult<IIdsCursorQueryResultDTO>> iterator,
            int maxPageSize);
    }

    public class MultiLevelCursorIteratorFactory : IMultiLevelCursorIteratorFactory
    {
        public IMultiLevelCursorIterator<TInput, TOutput> Create<TDTO, TInput, TOutput>(
            ITwitterPageIterator<ITwitterResult<TDTO>> pageIterator,
            Func<TDTO, TInput[]> transform,
            Func<TInput[], Task<TOutput[]>> getChildItems,
            int maxPageSize)
        {
            var iterator = new MultiLevelCursorIterator<TInput, TOutput>(
                async () =>
                {
                    var userIdsPage = await pageIterator.NextPageAsync().ConfigureAwait(false);

                    return new CursorPageResult<TInput, string>
                    {
                        Items = transform(userIdsPage.Content.Model),
                        NextCursor = userIdsPage.NextCursor,
                        IsLastPage = userIdsPage.IsLastPage
                    };
                }, async inputs =>
                {
                    var userIdsToAnalyze = inputs.Take(maxPageSize).ToArray();
                    var friends = await getChildItems(userIdsToAnalyze).ConfigureAwait(false);

                    return new MultiLevelPageProcessingResult<TInput, TOutput>
                    {
                        Items = friends,
                        AssociatedParentItems = userIdsToAnalyze,
                    };
                });

            return iterator;
        }

        public IMultiLevelCursorIterator<long, IUser> CreateUserMultiLevelIterator(
            ITwitterClient client,
            ITwitterPageIterator<ITwitterResult<IIdsCursorQueryResultDTO>> iterator,
            int maxPageSize)
        {
            return Create(iterator, dtoIds => dtoIds.Ids, client.Users.GetUsersAsync, maxPageSize);
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Iterators/MultiLevelCursorIteratorFactory.ts
import {ITwitterResult} from "../Web/TwitterResult";
import {ITwitterPageIterator} from "./TwitterPageIterator";
import {IMultiLevelCursorIterator} from "../../Public/Iterators/IMultiLevelCursorIterator";
import {ITwitterClient} from "../../Public/ITwitterClient";
import {IIdsCursorQueryResultDTO} from "../../Public/Models/Interfaces/DTO/QueryDTO/IIdsCursorQueryResultDTO";
import {IUser} from "../../Public/Models/Interfaces/IUser";
import {MultiLevelCursorIterator} from "./MultiLevelCursorIterator";
import {CursorPageResult} from "./Models/CursorPageResult";
import {MultiLevelPageProcessingResult} from "./Models/MultiLevelPageProcessingResult";
import {InjectionToken} from "@angular/core";

export interface IMultiLevelCursorIteratorFactory {
  create<TDTO, TInput, TOutput>(
    pageIterator: ITwitterPageIterator<ITwitterResult<TDTO>>,
    transform: (tDTO: TDTO) => TInput[],
    getChildItems: (tInput: TInput[]) => Promise<TOutput[]>,
    maxPageSize: number): IMultiLevelCursorIterator<TInput, TOutput>;

  createUserMultiLevelIterator(
    client: ITwitterClient,
    iterator: ITwitterPageIterator<ITwitterResult<IIdsCursorQueryResultDTO>>,
    maxPageSize: number): IMultiLevelCursorIterator<number, IUser>;
}

export class MultiLevelCursorIteratorFactory implements IMultiLevelCursorIteratorFactory {
  public create<TDTO, TInput, TOutput>(
    pageIterator: ITwitterPageIterator<ITwitterResult<TDTO>>,
    transform: (tDTO: TDTO) => TInput[],
    getChildItems: (tInput: TInput[]) => Promise<TOutput[]>,
    maxPageSize: number): IMultiLevelCursorIterator<TInput, TOutput> {
    let iterator: MultiLevelCursorIterator<TInput, TOutput> = new MultiLevelCursorIterator<TInput, TOutput>(async () => {
      let userIdsPage = await pageIterator.nextPageAsync(); // .ConfigureAwait(false);

      let result = new CursorPageResult<TInput, string>();
      result.items = transform(userIdsPage.result.Content.Model);
      result.nextCursor = userIdsPage.result.NextCursor;
      result.isLastPage = userIdsPage.result.IsLastPage;

      return result;
    }, async inputs => {
      let userIdsToAnalyze = inputs.filter((tInput, num) => num < maxPageSize); // Take(maxPageSize).ToArray();
      let friends = await getChildItems(userIdsToAnalyze); // .ConfigureAwait(false);

      let result = new MultiLevelPageProcessingResult<TInput, TOutput>();
      result.items = friends;
      result.associatedParentItems = userIdsToAnalyze;

      return result;
    });

    return iterator;
  }

  public createUserMultiLevelIterator(client: ITwitterClient, iterator: ITwitterPageIterator<ITwitterResult<IIdsCursorQueryResultDTO>>, maxPageSize: number): IMultiLevelCursorIterator<number, IUser> {
    return this.create(iterator, dtoIds => dtoIds.ids, client.users.getUsersAsync, maxPageSize);
  }
}


export const IMultiLevelCursorIteratorFactoryToken = new InjectionToken<IMultiLevelCursorIteratorFactory>('IMultiLevelCursorIteratorFactory', {
  providedIn: 'root',
  factory: () => new,
});
